405 creation editing entrance rigging

YohanHATOT opened the pull request.

YohanHATOT committed:
feat(#405): riggings edit and create - intern work

urien commented:
Actuellement je ne peux pas créer un rigging s'il n'y en a pas déjà un. Il faudrait afficher quelque chose qui ressemble à ce qui est utilisé pour les autres champs 
![image](https://user-images.githubusercontent.com/1165144/187698716-72224ca9-7fdc-4850-9367-221ca9de84f6.png)

urien commented:
J'ai édité une fiche d'équipement et quand j'ai validé, c'est la fiche d'équipement initiale qui s'affiche. il faudrait que ce soit la fiche d'équipement mise à jour

urien commented:
J'ai fait une mise à jour en supprimant une ligne. Cette ligne est bien supprimée à l'affichage mais quand on valide ce n'est pas le cas, elle existe toujours https://orange-rock-0d4f87503-735.westeurope.azurestaticapps.net/ui/entrances/37497

YohanHATOT commented:
> Actuellement je ne peux pas créer un rigging s'il n'y en a pas déjà un. Il faudrait afficher quelque chose qui ressemble à ce qui est utilisé pour les autres champs ![image](https://user-images.githubusercontent.com/1165144/187698716-72224ca9-7fdc-4850-9367-221ca9de84f6.png)

Nous avons résolu ce que vous nous avez dit de faire dans ce message, sauf qu'il manque les traductions pour les riggings et comentaires.

urien commented:
Pour les traductions il n'y a pas de pb si vous utilisez bien le fichier de langue en Anglais pour vos textes. Dès que votre code sera publié nous aurons une notification nous indiquant qu'il y a de nouveaux textes à traduire. Quand toutes les traductions d'une langue sont proposées Transifex fait une pull request

urien reviewed (changes requested):
Il faudrait, comme pour les autres pavés, disposer d'un + pour créer un rigging
![image](https://user-images.githubusercontent.com/1165144/190204090-51b01eac-a5b4-4c26-a33e-136504348e74.png)

La suppression d'une ligne dans un tableau existant ne fonctionne pas (ligne test à supprimer) 
https://orange-rock-0d4f87503-735.westeurope.azurestaticapps.net/ui/entrances/37497

clemsix6 commented:
Il y a actuellement un problème, lorsqu'on update les riggings le message d'erreur ci-dessous apparait dans la console, en rechange la requête est bien envoyée car lorsqu'on refresh la page les modifications sont bien apportés
![Screenshot_20220916_132817](https://user-images.githubusercontent.com/49559255/190630337-762027e9-e744-40de-9748-4405222bed33.png)

bsoufflet commented:
> Il y a actuellement un problème, lorsqu'on update les riggings le message d'erreur ci-dessous apparait dans la console, en rechange la requête est bien envoyée car lorsqu'on refresh la page les modifications sont bien apportés ![Screenshot_20220916_132817](https://user-images.githubusercontent.com/49559255/190630337-762027e9-e744-40de-9748-4405222bed33.png)

@clemsix6 Je pense que le problème vient du fait que l'objet envoyé à l'API pour un update (string séparé avec des `|;|`) n'est pas le meme que celui utilisé pour l'affichage des riggings (en mode view).
Dans le reducer tu passes l'objet Riggings que tu as envoyé en PATCH (ou POST dans le cas du create) et celui-ci est ensuite directement réutilisé pour l'affichage des Riggings en mode "Read" or l'objet en mode read n'a pas le meme format !!!

C'est dans le `EntranceReducer.js` ligne 137 et 145 que tu passes l'objet de l'update (ou du create) et qu'il n'est pas au bon format (je pense...).
Il y a 2 solutions, mais l'une bien qu'idéal demandera trop de temps  : 
- modifier la route de back pour que le format de Rigging soit le meme dans le cas du READ et de l'UPDATE/CREATE
- modifier l'objet dans le reducer pour le formater de la manière attendue par le composant Riggings en read only.

Je te laisse regarder cela :)

Tu en penses quoi @Clm-Roig ?

Clm-Roig commented:
Je suis d'accord avec toi, pour l'instant il faut juste formatter de la bonne manière pour le front. Tant pis si le format diffère entre update / create et read côté back. 

De toute façon, je pense qu'il faudra retravailler les riggings côté serveur et migrer les données pour ne plus avoir ce format avec séparateur qui n'est pas idéal 🙂

urien commented:
Je viens de faire un test d'édition d'un rigging existant, je n'ai pas de message d'erreur dans la console mais la mise à jour n'est pas réalisée.

urien commented:
Il manque encore le + qui permette de créer un nouveau Rigging

Clm-Roig commented on packages/web-app/src/reducers/EntranceReducer.js:
```diff
@@ -29,6 +31,30 @@ const initialState = {
+  const ropes = data.ropes.split('|;|');
+  const anchors = data.anchors.split('|;|');
+  const observations = data.observations.split('|;|');
+  const rigging = {
+    id: data.id,
+    language: data.language,
+    title: data.title,
+    obstacles: []
+  };
```
Il manque des infos à propos du rigging quand tu fais ça, notamment l'auteur du rigging. Ce qui conduit à "unknown author" après une édition de rigging. 

Je te conseille de faire la chose suivante ([destructuration](https://laravel.sillo.org/es6-la-destructuration/)) pour préserver un maximum de données : 
```js
const { obstacles, ropes, anchors, observations, ...riggingRest } = data;
const rigging = {
  ...riggingRest,
  id: data.id,
  language: data.language,
  title: data.title,
  obstacles: []
}
```

Comme ça tu exclus les "vieux" obstacles / ropes / anchors / observations et tu préserves toutes les autres données. Bonus : si des nouveaux champs apparaissent dans data, ils seront eux aussi préservés 😉

Clm-Roig commented on packages/web-app/src/components/common/Layouts/Fixed/ActionBar.jsx:
```diff
@@ -1,5 +1,5 @@
 import React, { useContext } from 'react';
-import { isNil, propOr } from 'ramda';
+import { isNil } from 'ramda';
```
Il y a un souci de rebase sur ce fichier / commit : il n'est pas censé apparaître dans cette PR...

Clm-Roig commented on packages/web-app/src/components/appli/Form/RiggingsForm/index.jsx:
```diff
@@ -0,0 +1,284 @@
+  };
+  const [v, setV] = useState(values.obstacles);
+
+  const addValue = () => {
+    const newV = [...v];
+    newV.push('');
+    setV(newV);
+  };
+
+  const rmvValue = index => {
+    const newV = [...v];
+    newV.splice(index, 1);
+    setV(newV);
```
Évite les acronymes sur les noms de variables : ça fait gagner quelques pixels mais c'est plus dur à relire et à maintenir pour les autres développeurs. Ce n'est pas grave si tu te retrouves avec des noms qui font 30 caractères : tant que c'est clair, c'est le principal ! 🙂 

rmv => remove
____V => ____Value

Clm-Roig commented on packages/web-app/src/components/appli/Form/RiggingsForm/index.jsx:
```diff
@@ -0,0 +1,284 @@
+}) => {
+  const { formatMessage } = useIntl();
+  const { languageObject } = useSelector(state => state.intl);
+  const MultilinesTableCell = styled(TableCell)({ whiteSpace: 'pre-wrap' });
```
À sortir du composant et à mettre au-dessus avec les autres sous-composants stylisés.

Clm-Roig reviewed (changes requested):
En plus de mes commentaires sur les fichiers, tu as commit un fichier vide au chemin suivant : packages/web-app/src/actions/UpdateRiggings.jsx~

Il faudrait le supprimer s'il te plaît 🙂

Clm-Roig reviewed (changes requested):
Cf mon précédent commentaire https://github.com/GrottoCenter/grottocenter-front/pull/735#pullrequestreview-1117326659

clemsix6 commented:
Je penses qu'il y a un probleme avec la fonction CreateRiggingsForm, je l'utilise comme CreateCommentForm sauf que les erreurs suivantes apparaissent et j'ai pas trouvé comment les résoudre
![image](https://user-images.githubusercontent.com/49559255/191868226-42733be1-7b61-4a6a-9b9d-51d05b90b786.png)

bsoufflet commented:
> CreateRiggingsForm

![image](https://user-images.githubusercontent.com/985903/191951389-f5255fa0-78a5-406c-84d5-e64c9903160c.png)
Solution

urien commented:
Je rencontre une difficulté. 
Quand je veux éditer ce rigging
![image](https://user-images.githubusercontent.com/1165144/192039617-e7176663-2b6b-4363-84b2-d47ee7f3f66d.png)
Je me retrouve face à un tableau vide. Ce qui était dedans a disparu
![image](https://user-images.githubusercontent.com/1165144/192039815-22f2da4d-f472-4391-bab4-380bccff7eb3.png)
Pour une cavité qui n'a pas de fiche d'équipement il n'y a pas de module vide permettant d'en créer une 
![image](https://user-images.githubusercontent.com/1165144/192040215-3eebd220-bacc-43ae-8e1d-fc88a8d198f1.png)

urien commented:
Maintenant il est possible de créer un rigging sur une cavité qui n'en a pas. Par contre le nom de l'auteur n'est pas enregistré
![image](https://user-images.githubusercontent.com/1165144/192043771-c99cceac-ef3f-4cc2-99bc-a255f6737479.png)
Lors de l'édition d'un rigging on peut modifier les textes mais il n'est pas possible de supprimer des lignes. cela s'affiche bien à l'édition mais ce n'est pas enregistré
C'est en bonne voie ;-)

bsoufflet committed:
feat(#405): create / edit rigging - fixes and improv

also :
- use the new API format for obstacles during update / create
- Allow empty body for contribution component
- order riggings list by title
- use same cancel button
- better handle empty rating on edit comments form (do not set it to 0)

bsoufflet commented:
@urien : J'ai fait des corrections du coté de l'API (PR ouverte). C'est donc normal que si tu testes ici, tu doives refresh la page d'une entrée après avoir modifié / ajouté un riggings.

urien reviewed (approved):
C'est remarquable comme fonctionnement !
Le fait de pouvoir changer l'ordre des lignes c'est royal
Il y a encore un petit pb de mise à jour après validation, mais cela va faire une très belle fonctionnalité

maximenathan commented on packages/web-app/src/components/appli/Entry/Riggings/Riggings.jsx:
```diff
@@ -0,0 +1,76 @@
+// eslint-disable-next-line react/prop-types
```
```suggestion
```

maximenathan commented on packages/web-app/src/components/appli/Entry/Riggings/Riggings.jsx:
```diff
@@ -0,0 +1,76 @@
+            <>
+              <CreateRiggingsForm
+                closeForm={() => setIsFormVisible(false)}
+                isNew={isNew}
```
```suggestion
                isNew
```

maximenathan commented on packages/web-app/src/components/appli/Entry/Riggings/Riggings.jsx:
```diff
@@ -0,0 +1,76 @@
+    dispatch(postRiggings(data, entranceId, data.language.id, riggings.id));
+    setIsFormVisible(false);
+  };
+  const isNew = true;
```
```suggestion

```

maximenathan commented on packages/web-app/src/components/appli/Entry/Riggings/Riggings.jsx:
```diff
@@ -0,0 +1,76 @@
+  const dispatch = useDispatch();
+  const permissions = usePermissions();
+  const [isFormVisible, setIsFormVisible] = useState(false);
+  const onSubmitForm = data => {
```
**Non-blocking:**

```suggestion
  const handleSubmitForm = data => {
```

maximenathan commented on packages/web-app/src/components/appli/Entry/Riggings/Riggings.jsx:
```diff
@@ -0,0 +1,76 @@
+          {riggings
+            ?.sort((r1, r2) => r1.title.localeCompare(r2.title))
+            .map((rigging, i) => (
+              <Rigging riggings={riggings} rigging={rigging} index={i} />
```
Blocking:
- use the unique id
- why passing "riggings"? Look like it's just for the divider. We could add the divider instead


Not tested:
```suggestion
            .map((rigging) => (
             <>
              <Rigging rigging={rigging} index={rigging.id} />
              {riggings.length - 1 && <Divider variant="middle" />}
             </>
```

maximenathan commented on packages/web-app/src/pages/Entry/transformers.js:
```diff
@@ -20,9 +20,10 @@ export const getComments = comments =>
         : null,
       id: comment?.id,
-      interestRate: pathOr(0, ['aestheticism'], comment) / 2,
+      // eslint-disable-next-line no-unsafe-optional-chaining
```
```suggestion

```

Souterweb commented:
Le 02/11/2022 à 00:35, github-actions[bot] a écrit :
>
> Azure Static Web Apps: Your stage site is ready! Visit it here: 
> https://orange-rock-0d4f87503-735.westeurope.azurestaticapps.net
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/GrottoCenter/grottocenter-front/pull/735#issuecomment-1299357518>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AE3PETOZOX454YJPKIYHHYTWGGSM5ANCNFSM6AAAAAAQBF6S2Y>.
> You are receiving this because you are subscribed to this 
> thread.Message ID: 
> ***@***.***>
>
  Ave,

Yes !!! J'ai testé sur ce lien et l'édition des fiches d'équipement 
fonctionne.
Je n'ai pas encore testé la création à partir de rien.
Beau boulot en tout cas.

Par contre y aurait-il un moyen de créer une "moulinette" pour remplacer 
sur l’ensemble de la base dans toutes les fiches d'équipement (et 
uniquement là) le caractère *\n* qui est venu se mettre à la place d'un 
saut de ligne (shift Entrée) au moment du passage à la V3 ?

Domi

urien commented:
Lorsque je créé une fiche d'équipement (sans obstacle) et que je valide la page se met bien à jour. Par contre lorsque je souhaite éditer la fiche d'équipement que je viens de créer je tombe sur une page blanche avec comme erreur
TypeError: e.map is not a function
Error: react-error-boundary requires either a fallback, fallbackRender, or FallbackComponent prop
Si je rafraichi la page tout est ok

Il faudrait homogénéiser les titres, celui des fiches d'équipement n'est pas comme celui des autres modules
![image](https://user-images.githubusercontent.com/1165144/199418670-dceea7e4-1d76-440a-a996-ca1bf4fe73f6.png)

bsoufflet commented:
> Par contre y aurait-il un moyen de créer une "moulinette" pour remplacer sur l’ensemble de la base dans toutes les fiches d'équipement (et uniquement là) le caractère *\n* qui est venu se mettre à la place d'un saut de ligne (shift Entrée) au moment du passage à la V3 ? Domi

@Souterweb : C"est fait et directement visible en production ! Tous les "\n" ont été remplacé par un vrai caractère de saut de ligne pour les riggings !!

Souterweb commented:
Le 02/11/2022 à 23:59, Benjamin a écrit :
>
>     Par contre y aurait-il un moyen de créer une "moulinette" pour
>     remplacer sur l’ensemble de la base dans toutes les fiches
>     d'équipement (et uniquement là) le caractère /\n/ qui est venu se
>     mettre à la place d'un saut de ligne (shift Entrée) au moment du
>     passage à la V3 ? Domi
>
> @Souterweb <https://github.com/Souterweb> : C"est fait et directement 
> visible en production ! Tous les "\n" ont été remplacé par un vrai 
> caractère de saut de ligne pour les riggings !!
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/GrottoCenter/grottocenter-front/pull/735#issuecomment-1301469928>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AE3PETNL26PEA3YKADR7PB3WGLW3RANCNFSM6AAAAAAQBF6S2Y>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
>

YESSSS !!!!

Super génial....

Merci à toi ô grand manitou des bits.

Domi

bsoufflet commented:
@urien 
> Lorsque je créé une fiche d'équipement (sans obstacle) et que je valide la page se met bien à jour. Par contre lorsque je souhaite éditer la fiche d'équipement que je viens de créer je tombe sur une page blanche avec comme erreur TypeError: e.map is not a function Error: react-error-boundary requires either a fallback, fallbackRender, or FallbackComponent prop Si je rafraichi la page tout est ok

Ce problème est lié à l'API qui n'est pas à jour et qui renvoie un mauvais résultat après une sauvegarde. avec la mise à jour de l'API ce sera ok.

> Il faudrait homogénéiser les titres, celui des fiches d'équipement n'est pas comme celui des autres modules ![image](https://user-images.githubusercontent.com/1165144/199418670-dceea7e4-1d76-440a-a996-ca1bf4fe73f6.png)

C'est fait. H4 pour tous les titres de cette page.

bsoufflet commented:
@urien : Je te laisse me confirmer si cela te convient maintenant ?

urien reviewed (approved):
C'est parfait.
Merci beaucoup
